src/java.desktop/share/classes/java/awt/AWTEventMulticaster.java

```diff
@@ -954,14 +954,75 @@ public static MouseWheelListener remove(MouseWheelListener l,
      * If listener-b is null, it returns listener-a
      * If neither are null, then it creates and returns
      * a new AWTEventMulticaster instance which chains a with b.
+     * <p>
+     * Occasionally (after approximately 500 invocations) the root
+     * AWTEventMulticaster is rebalanced.
+     *
      * @param a event listener-a
      * @param b event listener-b
      * @return the resulting listener
      */
     protected static EventListener addInternal(EventListener a, EventListener b) {
         if (a == null)  return b;
         if (b == null)  return a;
-        return new AWTEventMulticaster(a, b);
+        AWTEventMulticaster n = new AWTEventMulticaster(a, b);
+        if (!needsRebalance(n))
+            return n;
+
+        EventListener[] array = getListeners(n, EventListener.class);
+        return rebalance(array, 0, array.length - 1);
+    }
+
+    /**
+     * Return true if the argument represents a binary tree that needs to be rebalanced.
+     * <p>
+     * The criteria for when we need a rebalance is subjective. For now this method
+     * checks up to 500 of the topmost nodes of a AWTEventMulticaster. If they all include
+     * one leaf node, then this method returns true. This criteria will be met after
+     * 500 iterations of {@link #addInternal(EventListener, EventListener)}.
+     * <p>
+     * The first time this method is invoked it will convert a tree that has a degree
+     * of approximately 500 to a tree with a degree of approximately 10.
+     */
+    private static boolean needsRebalance(AWTEventMulticaster l) {
+        int level = 0;
+        while (true) {
+            if (++level > 500)
+                return true;
```

Comment:
@prrace : I added { .. }

@bourgesl : I'm struggling to define this as a constant in a helpful way. Is there an example of a similar constant somewhere in the codebase you could refer me to for comparison?

I think part of my challenge here is the AWTEventMulticaster doesn't promise to be any particular kind of tree. (For example: if it promised it was a self-balancing tree, then we'd know what our obligations were to meet that expectation.)

This PR's implementation is intended to be a fast heuristic/educated guess, but it does not scan the entire height of the left or right node. This makes it ... hard for me to name.

(Relatedly: the value 500 is arbitrary, and I'm happy to change it if someone would rather it be closer to 50 or 1000.)

Reply:
I added a constant `MAX_UNBALANCED_TOP_NODES = 100`, and I moved some comments from outside the method to inside the method.

Regarding recursion:

I'm reluctant to explore the tree recursively because I want `needsRebalance` to be a light/fast call. (It will be called with every addition.)

The primary use case I'm worried about is code that resembles this example in Container.java:
```
public synchronized void addContainerListener(ContainerListener l) {
        if (l == null) {
            return;
        }
        containerListener = AWTEventMulticaster.add(containerListener, l);
        newEventsOnly = true;
    }
```

There are several of these static `AWTEventMulticaster.add(X, X)` methods. This is (IMO) the recommended/primary way to interact with the AWTEventMulticaster.

This implementation (without rebalancing) created a tree where the *top* of the tree was guaranteed to be unbalanced. (Specifically: every left node may be a tree, and every right node will be a leaf.) It's not necessary to recursively scan the entire tree to identify this condition, so I'd prefer not to make `needsRebalance` more complex than it needs to be to identify this usage. (That is: there's no need to look at the middle or bottom of the tree, or even to identify the size or height of the tree.)

Regarding a System property:

I'm also reluctant to add this; it feels like bloat to me that nobody will ever use. Do you think you (or anyone else) will want to customize this property over time? My broad goal here is to prevent a StackOverflowError; not to fine-tune tree performance.

I definitely appreciate seeing `System.getProperty("x")` in several important pieces of code, but IMO altering the `MAX_UNBALANCED_TOP_NODES` field will only ever yield negligible changes. (Or that could be the subject of another ticket, with another test case?)

Reply:
@bourgesl before I try to integrate these changes: does my response above make sense/satisfy your concern?

Reply:
LGTM